### PR TITLE
JSON implementation

### PR DESCRIPTION
**Pull request by kustosz**
### Pull Request Description
Implements more generic, nicer and very awesome utilities for working with JSON.

### Important Notes
Performance is not great (but not terrible, for an interpreted language!) – the parsing pipeline is currently 10X slower than Scala's Circe. All of this slowdown is attributable to the cost of foreign calls across the polyglot boundary and the Enso interpreter itself, therefore further work towards improving it has been scheduled.

### Checklist

Please include the following checklist in your PR:

- [x] The documentation has been updated if necessary.
- [x] All code conforms to the [Scala](https://github.com/enso-org/enso/blob/main/docs/style-guide/scala.md), [Java](https://github.com/enso-org/enso/blob/main/docs/style-guide/java.md), and [Rust](https://github.com/enso-org/enso/blob/main/docs/style-guide/rust.md) style guides.
- [x] All documentation and configuration conforms to the [markdown](https://github.com/enso-org/enso/blob/main/docs/style-guide/markdown.md) and [YAML](https://github.com/enso-org/enso/blob/main/docs/style-guide/yaml.md) style guides.
- [x] All code has been tested where possible.

### CODE REVIEW

File: engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/mutable/GetRefNode.java [hunk 1]
```diff
@@ -0,0 +1,13 @@
+package org.enso.interpreter.node.expression.builtin.mutable;
+
+import com.oracle.truffle.api.nodes.Node;
+import org.enso.interpreter.dsl.BuiltinMethod;
+import org.enso.interpreter.runtime.data.Ref;
+
+@BuiltinMethod(type = "Ref", name = "new", description = "Creates an empty ref.")
```
Comment (radeusgd) [inline engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/mutable/GetRefNode.java]: Shouldn't this be?
```suggestion
@BuiltinMethod(type = "Ref", name = "get", description = "Gets current value stored in ref.")
```

File: engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/mutable/PutRefNode.java [hunk 2]
```diff
@@ -0,0 +1,15 @@
+package org.enso.interpreter.node.expression.builtin.mutable;
+
+import com.oracle.truffle.api.nodes.Node;
+import org.enso.interpreter.dsl.BuiltinMethod;
+import org.enso.interpreter.runtime.data.Ref;
+
+@BuiltinMethod(type = "Ref", name = "new", description = "Creates an empty ref.")
```
Comment (radeusgd) [inline engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/mutable/PutRefNode.java]: Same as above, but `put`.

File: std-bits/src/main/java/org/enso/base/json/Parser.java [hunk 3]
```diff
@@ -0,0 +1,449 @@
...
+        return position + 5;
+      }
+      throw new UnexpectedTokenException(position, "a false");
+    }
+    throw new UnexpectedEndOfInputException("a false");
+  }
+
+  private static int consumeString(
+      char[] chars, int position, JsonConsumer consumer, boolean isKey) {
+    if (chars[position] != '\"') {
```
Comment (radeusgd) [inline std-bits/src/main/java/org/enso/base/json/Parser.java]: Here you escape the `"` inside of char quotes `' '` but a few lines below you do not.

File: test/Benchmarks/src/Json.enso [hunk 4]
```diff
@@ -0,0 +1,13 @@
+from Base import all
+
+import Base.Bench_Utils
+
+prep_json size =
+    single = '{"foo": 543}, {"bar": false}'
+    many = Vector.new size (_ -> single)
+    str = "[" + (many.join ",") + "]"
+    str
```
Comment (radeusgd) [inline test/Benchmarks/src/Json.enso:9]: Just a suggestion: Maybe the benchmark could also test deeper structures?

File: distribution/std-lib/Base/src/Data/Json/Internal.enso [hunk 5]
```diff
@@ -0,0 +1,170 @@
...
+        String v -> v
+        _ -> Panic.throw (Type_Mismatch_Error json fmt)
+    _ ->
+        m = Meta.meta fmt
+        case m of
+            Meta.Atom _ -> case json of
+                Object json_fields ->
+                    cons = Meta.Constructor m.constructor
+                    fnames = cons . fields
+                    ffmts = m.fields
```
Comment (radeusgd) [inline distribution/std-lib/Base/src/Data/Json/Internal.enso]: Why the first line has extra whitespace but the second one doesn't?

**Review by radeusgd — APPROVED** [review]
Looks mostly good to me, just a few comments.

I wonder why the CI did not trigger.

File: distribution/std-lib/Base/src/Main.enso [hunk 6]
```diff
@@ -95,3 +97,4 @@ Any.== that = if Meta.is_same_object this that then True else
            Therefore, there is no more cases to handle in this method.
         _ -> False
 
+Boolean.to_json = Json.Boolean this
```
Comment (iamrecursion) [inline distribution/std-lib/Base/src/Main.enso:102]: This should probably go in a `Data.Boolean.Extensions` module like we have for `Number` and `Text`.

File: engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/mutable/NewNode.java [hunk 7]
```diff
@@ -1,4 +1,4 @@
-package org.enso.interpreter.node.expression.builtin.array;
+package org.enso.interpreter.node.expression.builtin.mutable;
```
Comment (iamrecursion) [inline engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/mutable/NewNode.java:1]: I still really think these `NewNode` nodes should be done using in-language variadics. :P

File: engine/runtime/src/main/java/org/enso/interpreter/runtime/builtin/Builtins.java [hunk 8]
```diff
@@ -235,8 +233,8 @@ public System system() {
   }
 
   /** @return the container for array-related builtins. */
-  public Array array() {
-    return array;
+  public Mutable array() {
```
Comment (iamrecursion) [inline engine/runtime/src/main/java/org/enso/interpreter/runtime/builtin/Builtins.java]: Surely this call should be renamed to `mutable`?

File: std-bits/src/main/java/org/enso/base/json/Parser.java [hunk 9]
```diff
@@ -0,0 +1,449 @@
...
+    throw new UnexpectedTokenException(position, "a start of a JSON value");
+  }
+
+  private static int consumeNull(char[] chars, int position, JsonConsumer consumer) {
+    if (position + 3 < chars.length) {
+      boolean match =
+          chars[position] == 'n'
+              && chars[position + 1] == 'u'
+              && chars[position + 2] == 'l'
+              && chars[position + 3] == 'l';
```
Comment (iamrecursion) [inline std-bits/src/main/java/org/enso/base/json/Parser.java:240]: Well this is gross :'D

**Review by iamrecursion — APPROVED** [review]

**Pull request merged**